crates/valence_new/src/packet_stream.rs

```diff
@@ -0,0 +1,181 @@
+use std::collections::VecDeque;
+
+use bytes::{BufMut, BytesMut};
+use valence_protocol::packets::S2cPlayPacket;
+use valence_protocol::{DecodePacket, EncodePacket, PacketDecoder, PacketEncoder};
+
+use crate::server::byte_channel::{ByteReceiver, ByteSender, TryRecvError};
+
+/// Represents a byte stream that packets can be read from and written to.
+pub trait PacketStream {
```

Comment:
`try_send` and `try_recv` need to operate on bytes of encoded packets rather than actual packet objects. `PacketEncoder` and `PacketDecoder` should stay in `Client`. It's meant to represent the interface that a pair of byte channels gives you. We couldn't do it this way even if we wanted because the trait is not object-safe and we are putting it in a `Box<dyn PacketStream>`.

(I didn't really make this clear sorry)

Reply:
Yeah I just discovered this :(